File: api/update.js
/**
 * Update Endpoint
 * Triggar inkrementell scraping med säker autentisering
 */

const crypto = require('crypto');
const { exec } = require('child_process');
const { promisify } = require('util');
const execAsync = promisify(exec);

// Rate limiter (in-memory för enkelhetens skull)
const rateLimits = new Map();
const updateLocks = new Map();

function getClientIp(req) {
    const forwarded = req.headers['x-forwarded-for'];
    if (forwarded) {
        return forwarded.split(',')[0].trim();
    }
    return req.headers['x-real-ip'] || 'unknown';
}

function verifyToken(token, clientIp) {
    if (!token) {
        return { valid: false, message: 'Token saknas' };
    }
    
    try {
        const parts = token.split(':');
        if (parts.length !== 2) {
            return { valid: false, message: 'Ogiltigt token-format' };
        }
        
        const [timestampStr, providedHash] = parts;
        const timestamp = parseInt(timestampStr);
        
        // Check token age (max 1 timme)
        const now = Math.floor(Date.now() / 1000);
        const ageSeconds = now - timestamp;
        
        if (ageSeconds > 3600) {
            return { valid: false, message: 'Token har gått ut. Ladda om sidan.' };
        }
        
        if (ageSeconds < 0) {
            return { valid: false, message: 'Ogiltigt token' };
        }
        
        // Verifiera HMAC
        const secret = process.env.UPDATE_API_KEY || 'dev-key-change-in-production';
        const message = `${clientIp}:${timestamp}`;
        const hmac = crypto.createHmac('sha256', secret);
        hmac.update(message);
        const expectedHash = hmac.digest('hex');
        
        if (crypto.timingSafeEqual(Buffer.from(providedHash), Buffer.from(expectedHash))) {
            return { valid: true, message: 'Token verified' };
        }
        
        return { valid: false, message: 'Ogiltigt token' };
        
    } catch (error) {
        return { valid: false, message: `Token verification error: ${error.message}` };
    }
}

function checkRateLimit(clientIp) {
    const maxRequests = parseInt(process.env.MAX_UPDATES_PER_HOUR || '3');
    const cooldownMinutes = parseInt(process.env.COOLDOWN_MINUTES || '5');
    const cooldownSeconds = cooldownMinutes * 60;
    const windowSeconds = 3600; // 1 timme
    
    const now = Date.now() / 1000;
    const clientKey = crypto.createHash('sha256').update(clientIp).digest('hex').substring(0, 16);
    
    if (!rateLimits.has(clientKey)) {
        rateLimits.set(clientKey, []);
    }
    
    const requests = rateLimits.get(clientKey);
    
    // Rensa gamla requests
    const validRequests = requests.filter(ts => now - ts < windowSeconds);
    rateLimits.set(clientKey, validRequests);
    
    // Kolla cooldown
    if (validRequests.length > 0) {
        const lastRequest = Math.max(...validRequests);
        const timeSinceLast = now - lastRequest;
        
        if (timeSinceLast < cooldownSeconds) {
            const waitMinutes = Math.ceil((cooldownSeconds - timeSinceLast) / 60);
            return {
                allowed: false,
                message: `Vänta ${waitMinutes} minuter innan nästa uppdatering`
            };
        }
    }
    
    // Kolla max requests
    if (validRequests.length >= maxRequests) {
        const oldestRequest = Math.min(...validRequests);
        const timeUntilReset = Math.ceil((windowSeconds - (now - oldestRequest)) / 60);
        return {
            allowed: false,
            message: `För många uppdateringar. Försök igen om ${timeUntilReset} minuter`
        };
    }
    
    // Tillåt och lägg till timestamp
    validRequests.push(now);
    rateLimits.set(clientKey, validRequests);
    
    return { allowed: true, message: 'OK' };
}

function checkUpdateLock() {
    const now = Date.now() / 1000;
    
    // Cleanup gamla locks (timeout efter 10 minuter)
    for (const [key, lock] of updateLocks.entries()) {
        if (now - lock.timestamp > 600) {
            updateLocks.delete(key);
        }
    }
    
    // Kolla om något lock finns
    if (updateLocks.size > 0) {
        const lock = Array.from(updateLocks.values())[0];
        return {
            locked: true,
            message: `En uppdatering pågår redan (startad ${new Date(lock.timestamp * 1000).toLocaleTimeString('sv-SE')})`
        };
    }
    
    return { locked: false };
}

function acquireLock(clientId) {
    const lockCheck = checkUpdateLock();
    
    if (lockCheck.locked) {
        return { acquired: false, message: lockCheck.message };
    }
    
    updateLocks.set(clientId, {
        timestamp: Date.now() / 1000,
        clientId: clientId
    });
    
    return { acquired: true };
}

function releaseLock(clientId) {
    updateLocks.delete(clientId);
}

module.exports = async (req, res) => {
    // CORS headers
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type, X-Update-Token');
    
    // Handle OPTIONS
    if (req.method === 'OPTIONS') {
        res.status(200).end();
        return;
    }
    
    const clientIp = getClientIp(req);
    const clientId = crypto.createHash('sha256').update(clientIp).digest('hex').substring(0, 16);
    
    try {
        // 1. Verifiera token
        const token = req.headers['x-update-token'];
        const tokenCheck = verifyToken(token, clientIp);
        
        if (!tokenCheck.valid) {
            return res.status(401).json({
                success: false,
                message: tokenCheck.message,
                code: 'UNAUTHORIZED'
            });
        }
        
        // 2. Rate limiting
        const rateCheck = checkRateLimit(clientIp);
        if (!rateCheck.allowed) {
            return res.status(429).json({
                success: false,
                message: rateCheck.message,
                code: 'RATE_LIMITED'
            });
        }
        
        // 3. Check lock
        const lockCheck = checkUpdateLock();
        if (lockCheck.locked) {
            return res.status(409).json({
                success: false,
                message: lockCheck.message,
                code: 'LOCKED'
            });
        }
        
        // 4. Acquire lock
        const lock = acquireLock(clientId);
        if (!lock.acquired) {
            return res.status(409).json({
                success: false,
                message: lock.message,
                code: 'LOCKED'
            });
        }
        
        try {
            // 5. Kör scraper (Python)
            const { stdout, stderr } = await execAsync(
                'python3 scraper_incremental.py --incremental',
                { 
                    timeout: 300000, // 5 minuter
                    maxBuffer: 10 * 1024 * 1024 // 10MB
                }
            );
            
            // Parse output för ändringar
            const changes = {
                new: 0,
                updated: 0,
                deleted: 0
            };
            
            const newMatch = stdout.match(/(\d+)\s+nya/);
            const updatedMatch = stdout.match(/(\d+)\s+uppdaterade/);
            const deletedMatch = stdout.match(/(\d+)\s+raderade/);
            
            if (newMatch) changes.new = parseInt(newMatch[1]);
            if (updatedMatch) changes.updated = parseInt(updatedMatch[1]);
            if (deletedMatch) changes.deleted = parseInt(deletedMatch[1]);
            
            return res.status(200).json({
                success: true,
                message: 'Dokumentation uppdaterad',
                changes: changes,
                timestamp: new Date().toISOString()
            });
            
        } catch (error) {
            console.error('Scraper error:', error);
            
            return res.status(500).json({
                success: false,
                message: 'Fel vid uppdatering',
                error: process.env.DEBUG === 'true' ? error.message : 'Se serverloggar'
            });
            
        } finally {
            // Release lock
            releaseLock(clientId);
        }
        
    } catch (error) {
        console.error('Update endpoint error:', error);
        
        return res.status(500).json({
            success: false,
            message: `Serverfel: ${error.message}`,
            code: 'SERVER_ERROR'
        });
    }
};

File: api/update.py
"""
API endpoint för att trigga dokumentationsuppdatering
Körs som Vercel Serverless Function

Säkerhetsfunktioner:
- Rate limiting (max 3 updates/timme)
- Queue system (endast en update åt gången)
- API-nyckel autentisering
- Audit logging
"""

from http.server import BaseHTTPRequestHandler
import json
import subprocess
import os
from datetime import datetime
import hashlib

# Import våra säkerhetsmoduler
try:
    from .rate_limiter import UpdateRateLimiter
    from .update_queue import UpdateQueue
    from .auth import SecureAuth
    from .config import Config
except:
    # Fallback för lokal testning
    import sys
    sys.path.insert(0, os.path.dirname(__file__))
    from rate_limiter import UpdateRateLimiter
    from update_queue import UpdateQueue
    from auth import SecureAuth
    from config import Config

# Initialize säkerhetsmoduler
rate_limiter = UpdateRateLimiter(
    max_requests=Config.MAX_UPDATES_PER_HOUR,
    window_minutes=60,
    cooldown_minutes=Config.COOLDOWN_MINUTES
)
update_queue = UpdateQueue()
secure_auth = SecureAuth()

def get_client_ip(request):
    """Hämta klientens IP-adress"""
    # Vercel sätter X-Forwarded-For
    ip = request.headers.get('X-Forwarded-For', '').split(',')[0].strip()
    if not ip:
        ip = request.headers.get('X-Real-IP', 'unknown')
    return ip

def get_client_id(ip):
    """Hash IP för anonymisering i logs"""
    return hashlib.sha256(ip.encode()).hexdigest()[:16]

def verify_request(request):
    """
    Säker verifiering med flera lager
    
    1. IP-whitelist (om konfigurerad)
    2. Token-baserad auth (HMAC, ingen synlig nyckel)
    3. Eller session-baserad auth (för admin-panel)
    
    Returns:
        (bool, str, str): (authorized, reason, client_id)
    """
    client_ip = get_client_ip(request)
    client_id = get_client_id(client_ip)
    
    # Layer 1: IP Whitelist (om konfigurerad)
    ip_allowed, ip_message = secure_auth.verify_ip(client_ip)
    if not ip_allowed:
        return False, ip_message, client_id
    
    # Layer 2: Token-baserad auth
    update_token = request.headers.get('X-Update-Token')
    if update_token:
        token_valid, token_message = secure_auth.verify_update_token(update_token, client_ip)
        if token_valid:
            return True, "Token verified", client_id
        else:
            return False, token_message, client_id
    
    # Layer 3: Session-baserad auth (för admin)
    session_token = request.headers.get('X-Session-Token')
    if session_token:
        session_valid, session_message = secure_auth.verify_session(session_token)
        if session_valid:
            return True, "Session verified", client_id
        else:
            return False, session_message, client_id
    
    # Ingen giltig auth
    return False, "Ingen giltig autentisering (token eller session saknas)", client_id

def log_update_attempt(client_id, success, message, changes=None):
    """Logga update-försök för audit"""
    log_file = Path('/tmp/update_audit.log')
    
    log_entry = {
        'timestamp': datetime.now().isoformat(),
        'client_id': client_id,
        'success': success,
        'message': message,
        'changes': changes
    }
    
    try:
        with open(log_file, 'a') as f:
            f.write(json.dumps(log_entry) + '\n')
    except:
        pass

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        """Trigga en inkrementell scraping"""
        client_ip = get_client_ip(self)
        client_id = get_client_id(client_ip)
        
        try:
            # 1. Säker autentisering (IP + Token)
            authorized, auth_message, _ = verify_request(self)
            
            if not authorized:
                self.send_response(401)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                
                response = {
                    'success': False,
                    'message': auth_message,
                    'code': 'UNAUTHORIZED'
                }
                
                log_update_attempt(client_id, False, f'Unauthorized: {auth_message}')
                self.wfile.write(json.dumps(response).encode())
                return
            
            # 2. Kontrollera rate limiting
            allowed, rate_message = rate_limiter.is_allowed(client_id)
            if not allowed:
                self.send_response(429)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.send_header('Retry-After', '300')  # 5 minutes
                self.end_headers()
                
                response = {
                    'success': False,
                    'message': rate_message,
                    'code': 'RATE_LIMITED',
                    'stats': rate_limiter.get_stats(client_id)
                }
                
                log_update_attempt(client_id, False, rate_message)
                self.wfile.write(json.dumps(response).encode())
                return
            
            # 3. Kontrollera om update redan körs
            is_running, lock_info = update_queue.is_update_running()
            if is_running:
                # Lägg till i kö
                position = update_queue.add_to_queue(client_id)
                
                self.send_response(202)  # Accepted
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                
                response = {
                    'success': False,
                    'message': f'En uppdatering pågår redan. Du är nummer {position} i kön.',
                    'code': 'QUEUED',
                    'queue_position': position,
                    'queue_status': update_queue.get_queue_status()
                }
                
                self.wfile.write(json.dumps(response).encode())
                return
            
            # 4. Acquire lock
            acquired, lock_message = update_queue.acquire_lock(client_id)
            if not acquired:
                self.send_response(409)  # Conflict
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                
                response = {
                    'success': False,
                    'message': lock_message,
                    'code': 'LOCKED'
                }
                
                self.wfile.write(json.dumps(response).encode())
                return
            
            # CORS headers
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
            self.send_header('Access-Control-Allow-Headers', 'Content-Type, X-API-Key, Authorization')
            self.end_headers()
            
            # 5. Kör scraper i inkrementellt läge
            try:
                result = subprocess.run(
                    ['python3', 'scraper_incremental.py', '--incremental'],
                    capture_output=True,
                    text=True,
                    timeout=300,  # 5 minuter timeout
                    cwd=os.path.dirname(os.path.dirname(__file__))
                )
                
                if result.returncode == 0:
                    # Parse output för att hitta antal ändringar
                    output = result.stdout
                    changes = {
                        'new': 0,
                        'updated': 0,
                        'deleted': 0
                    }
                    
                    # Leta efter mönster som "✓ 5 nya artiklar", "✓ 3 uppdaterade", etc.
                    import re
                    new_match = re.search(r'(\d+)\s+nya', output)
                    updated_match = re.search(r'(\d+)\s+uppdaterade', output)
                    deleted_match = re.search(r'(\d+)\s+raderade', output)
                    
                    if new_match:
                        changes['new'] = int(new_match.group(1))
                    if updated_match:
                        changes['updated'] = int(updated_match.group(1))
                    if deleted_match:
                        changes['deleted'] = int(deleted_match.group(1))
                    
                    # Log success
                    log_update_attempt(client_id, True, 'Update completed', changes)
                    
                    response = {
                        'success': True,
                        'message': 'Dokumentation uppdaterad',
                        'timestamp': datetime.now().isoformat(),
                        'changes': changes,
                        'output': output if os.getenv('DEBUG') else None
                    }
                else:
                    # Log failure
                    log_update_attempt(client_id, False, f'Scraper failed: {result.stderr}')
                    
                    response = {
                        'success': False,
                        'message': 'Fel vid uppdatering',
                        'error': result.stderr if os.getenv('DEBUG') else 'Se serverloggar'
                    }
                
            finally:
                # 6. Release lock oavsett resultat
                update_queue.release_lock()
            
            self.wfile.write(json.dumps(response).encode())
            
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            
            response = {
                'success': False,
                'message': 'Serverfel',
                'error': str(e)
            }
            
            self.wfile.write(json.dumps(response).encode())
    
    def do_OPTIONS(self):
        """Handle CORS preflight"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

